mm2src/coins/eth/eth_swap_v2.rs

```diff
@@ -96,14 +125,358 @@ impl EthCoin {
+
+        self.sign_and_send_transaction(
+            0.into(),
+            Action::Call(taker_swap_v2_contract),
+            data,
+            U256::from(gas_limit),
+        )
+        .compat()
+        .await
+    }
+
+    /// Checks that taker payment state is `TakerApproved`.
+    /// Accepts a taker-approved payment transaction and returns it if the state is correct.
+    pub(crate) async fn search_for_taker_funding_spend_impl(
+        &self,
+        tx: &SignedEthTx,
+    ) -> Result<Option<FundingTxSpend<Self>>, SearchForFundingSpendErr> {
+        let (decoded, taker_swap_v2_contract) = self
+            .get_decoded_and_swap_contract(tx, TAKER_PAYMENT_APPROVE)
+            .await
+            .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("{}", e)))?;
+        let taker_status = self
+            .payment_status_v2(
+                taker_swap_v2_contract,
+                decoded[0].clone(), // id from takerPaymentApprove
+                &TAKER_SWAP_V2,
+                EthPaymentType::TakerPayments,
+                3,
+            )
+            .await
+            .map_err(|e| SearchForFundingSpendErr::Internal(ERRL!("{}", e)))?;
+        if taker_status == U256::from(TakerPaymentStateV2::TakerApproved as u8) {
+            return Ok(Some(FundingTxSpend::TransferredToTakerPayment(tx.clone())));
         }
+        Ok(None)
     }
 
+    /// Taker swap contract `spendTakerPayment` method is called for EVM based chains.
+    /// Returns maker spent payment transaction.
+    pub(crate) async fn sign_and_broadcast_taker_payment_spend_impl(
+        &self,
+        gen_args: &GenTakerPaymentSpendArgs<'_, Self>,
+        secret: &[u8],
+    ) -> Result<SignedEthTx, TransactionErr> {
+        // TODO need new consts and params for v2 calls
+        let gas_limit = match self.coin_type {
+            EthCoinType::Eth => U256::from(self.gas_limit.eth_receiver_spend),
+            EthCoinType::Erc20 { .. } => U256::from(self.gas_limit.erc20_receiver_spend),
+            EthCoinType::Nft { .. } => {
+                return Err(TransactionErr::ProtocolNotSupported(ERRL!(
+                    "NFT protocol is not supported for ETH and ERC20 Swaps"
+                )))
+            },
+        };
+        let (taker_swap_v2_contract, approve_func, token_address) = self
+            .taker_swap_v2_details(TAKER_PAYMENT_APPROVE, TAKER_PAYMENT_APPROVE)
+            .await?;
+        // Note: `TakerApproved` status was checked in `search_for_taker_funding_spend` function
+        let data = {
+            let decoded = try_tx_s!(decode_contract_call(approve_func, gen_args.taker_tx.unsigned().data()));
+            try_tx_s!(
+                self.prepare_spend_taker_payment_data(gen_args, secret, decoded, token_address)
+                    .await
+            )
+        };
+
+        let spend_payment_tx = self
+            .sign_and_send_transaction(0.into(), Action::Call(taker_swap_v2_contract), data, gas_limit)
```

Comment:
Why is it always `0` value is used with `sign_and_send_transaction` calls?

Reply:
we dont send value directly in that transactions. we do it only for `ethTakerPayment` call https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapTakerV2.sol#L63

for `spendTakerPayment` the actual amount to be spent is one of the parameters encoded in the transaction input data.
https://github.com/KomodoPlatform/etomic-swap/blob/5e15641cbf41766cd5b37b4d71842c270773f788/contracts/EtomicSwapTakerV2.sol#L166

Reply:
Can you create a constant for `0` and document it in the source code?

Reply:
> Can you create a constant for 0 and document it in the source code?

Done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2169/commits/d9e3c4ead8a166e571c28303f2d2d238f68b846f